src/snowflake/cli/api/metrics.py

```diff
@@ -86,7 +181,66 @@ def increment_counter(self, name: str, value: int = 1) -> None:
         else:
             self._counters[name] += value
 
+    @property
+    def current_span(self) -> Optional[CLIMetricsSpan]:
+        return self._in_progress_spans[-1] if len(self._in_progress_spans) > 0 else None
+
+    @contextmanager
+    def start_span(self, name: str) -> Iterator[CLIMetricsSpan]:
+        """
+        Starts a new span that tracks various metrics throughout its execution
+
+        Assumes that parent steps contain the entirety of their child steps
+        Parent steps are automatically populated as the most recently executed step, like a stack
+
+        Only one span can be active at a time; once one span finishes its parent becomes the active span
+
+        Spans are not emitted in telemetry if depth/total limits are exceeded
+
+        :raises CliMetricsInvalidUsageError: if the step name is empty
+        """
+        new_step = CLIMetricsSpan(
+            name=name,
+            start_time=time.monotonic() - self._monotonic_start,
+            parent=self.current_span,
```

Comment:
do we want to allow passing in a custom parent? that would allow us to trace parallel operations:
```python
@metrics.start_span(name="put_file", parent=metrics.current_span)
def _put_file(path: Path):
    ...

with concurrent.futures.ThreadPoolExecutor() as executor:
    for file in files_to_upload:
        executor.submit(_put_file, file)
```

which would give us a span structure like
```
+------------------------------------+
| root                               |
+------------------------------------+
  +--------------------------+
  | parent                   |
  +--------------------------+
    +-----------------+
    | put_file        |
    +-----------------+
    +-------------+
    | put_file    |
    +-------------+
    +---------------+
    | put_file      |
    +---------------+  
```

Reply:
I added support for this, had to add a contextvar in the class to make sure that `metrics.current_span` is thread local, but I couldn't find a way to make it so that it is automatically propagated between threads

E.g. for the below code, assertion 1 will fail but assertion 2 will fail:
```python
    with metrics.start_span("parent") as parent_span:

        def parallel_operation(seq_num: int):
            assert metrics.current_span is parent_span # ASSERTION 1
            with metrics.start_span("child", parent=parent_span) as local_span:
                assert metrics.current_span is local_span # ASSERTION 2

        executor = concurrent.futures.ThreadPoolExecutor()
        futures = [executor.submit(parallel_operation, seq_num=i) for i in range(3)]
```

Do you know if there is a way around this or are we ok with declaring this as undefined behaviour?

Reply:
discussed offline, let's hold off on this to avoid complexity for now. we can revisit later if necessary